scripts: support compile_commands.json in gen_app_partitions.py

tejlmand opened the pull request:
Fixes: #40590

This commit updates gen_app_partitions.py to include only files present
in the current build by extracting the information from the CMake
generated `compile_commands.json` file.

This ensures that object files in sub-projects, such as `empty_cpu0`,
will not be considered by the script.

Using the compile_commands.json instead of walking the whole build tree
for finding object files also improves performance:

Time of executing `gen_app_partitions.py` (Old):
__________________________
Executed in  480.06 millis
   usr time  425.83 millis
   sys time   49.55 millis

Time of executing `gen_app_partitions.py` (New):
________________________________________________________
Executed in   76.22 millis
   usr time   49.00 millis
   sys time   24.59 millis

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
scripts: support compile_commands.json in gen_app_partitions.py

Fixes: #40590

This commit updates gen_app_partitions.py to include only files present
in the current build by extracting the information from the CMake
generated `compile_commands.json` file.

This ensures that object files in sub-projects, such as `empty_cpu0`,
will not be considered by the script.

Using the compile_commands.json instead of walking the whole build tree
for finding object files also improves performance:

Time of executing `gen_app_partitions.py` (Old):
__________________________
Executed in  480.06 millis
   usr time  425.83 millis
   sys time   49.55 millis

Time of executing `gen_app_partitions.py` (New):
________________________________________________________
Executed in   76.22 millis
   usr time   49.00 millis
   sys time   24.59 millis

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

dcpleung commented:
Some concerns:
* This requires all supported toolchains to use `-o` to indicate output file. Visual Studio uses `/` instead of `-` for command line arguments (though we don't support it currently). I think all of our currently supported toolchains uses `-o`, so this may not be an issue.
* This only works for compiled output. If someone decides to include a prebuilt library, this will skip the partitions defined in that library.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -558,6 +558,14 @@ set(SOC_SERIES ${CONFIG_SOC_SERIES})
+    "Export CMake compile commands. Used by gen_app_partitions.py script"
+    FORCE
+)
+
```
Checked the overhead for this.

Was unable to measure a time overhead, and the file generated was 244k.

So generating this unconditionally is OK.

SebastianBoe replied:
Looks like we can now get rid of some code in twister.

/home/sebo/ncs/zephyr/scripts/pylib/twister/twisterlib.py
2045:            cmake_args.append("-DCMAKE_EXPORT_COMPILE_COMMANDS=1")

tejlmand replied:
> Was unable to measure a time overhead, and the file generated was 244k.
>
> So generating this unconditionally is OK.

Thanks for checking.
Initial version was actually doing this conditionally, but feedback on Discord suggested to do this unconditionally.

So great you checked the impact.

tejlmand replied:
Done, `CMAKE_EXPORT_COMPILE_COMMANDS` removed in twister.

tejlmand commented:
> - This requires all supported toolchains to use -o to indicate output file. Visual Studio uses / instead of - for command line arguments (though we don't support it currently). I think all of our currently supported toolchains uses -o, so this may not be an issue.

Not to mention, that the snippet generated is `ld` based, meaning it's supported by GNU ld, and thus only toolchains that understand that format. afaik Visual Code built-in linker doesn't support ld files, so really doesn't matter that Visual Studio uses `/` instead of `-`.

> This only works for compiled output. If someone decides to include a prebuilt library, this will skip the partitions defined in that library.

Exactly as current solution. Only object files are considered, not prebuilt libraries.
Notice this in existing solution:
https://github.com/zephyrproject-rtos/zephyr/blob/f35cdfc1e4d3a59541bffcaefea739def7cd5468/scripts/gen_app_partitions.py#L143-L145

Not to mention that `ELFFile()` only supports elf files, that is either final elf or object files, not library archives.

tejlmand committed:
twister: remove CMAKE_EXPORT_COMPILE_COMMANDS=1

The -DCMAKE_EXPORT_COMPILE_COMMANDS=1 is removed from twisterlib.py as
it is now always set by the Zephyr build system.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

SebastianBoe commented on scripts/gen_app_partitions.py:
```diff
@@ -237,6 +260,8 @@ def main():
 
     if args.directory is not None:
         parse_obj_files(partitions)
```
Do we need to retain the parse_obj_files function?

tejlmand replied:
Not for Zephyr use-case afaik, but i'm unsure if anyone downstream is using this script also, so I didn't wanted to remove it straight away.

SebastianBoe commented on scripts/gen_app_partitions.py:
```diff
@@ -149,6 +150,26 @@ def parse_obj_files(partitions):
+            compile_command = command.get('command')
+            compile_arg = object_pattern.search(compile_command)
+            obj_file = None if compile_arg is None else compile_arg.group(1)
+            if obj_file is not None:
```
IIRC

if obj_file:

is more pythonic

tejlmand replied:
fixed.
